### ISSUE TITLE
support separate pipeline for each namespace

### ISSUE THREAD

**Issue by yaliqin** [kubeflow/pipelines#4197]
EDIT from @Bobgy: This issue got 47 upvotes when requesting user feedback: https://github.com/kubeflow/pipelines/issues/1223#issuecomment-656508302.

### Proposed change
Currently with 1.0.2 version, Kubeflow Deployment with kfctl_k8s_istio shares pipelines for all namespaces defined in user profile. Separate pipeline support for each namespace is needed because the multi-user notebook server separation is already supported. It is natural to support pipeline separation.  

### Alternative options
NA

### Who would use this feature?
A lot of enterprises will benefit from this feature as allocating different namespaces to different teams is a common practice in Kubernetes and resources in existing namespaces can be effectively used.

### Suggest a solution
NA

**Comment by Bobgy** [comment]
FYI, in existing to be released KF 1.1, pipeline runs are already in user namespaces.

but the static pipeline yaml files are not separated. Do you want this additionally?

**Comment by Bobgy** [comment]
Related to https://github.com/kubeflow/pipelines/issues/1223

**Comment by Bobgy** [comment]
There are more clarification in https://github.com/kubeflow/pipelines/issues/1223#issuecomment-656507073.

If you agree with the proposal in https://github.com/kubeflow/pipelines/issues/1223#issuecomment-656508302, can you rephrase your description to just focus on pipeline resource for this issue.

**Comment by yaliqin** [comment]
I think the pipeline yaml files should be supported also for E2E separation.  Thanks. Already upvote in #1223 .

**Comment by Bobgy** [comment]
Sorry, I'm kind of oversubscribed because of extra Kubeflow duties recently, so I may not be able to take this issue soon.

Leaving my previous thoughts about this:

For use-case context: some users want full separation of pipelines, and some want no separation. The distinction is mostly organization culture and I think both requests are valid.

So I think a MVP UX I can imagine that satisfies both is to:
1. pipelines in DB should have a namespace column
    * if namespace == '' (empty string), then it's in the shared space
    * if namespace is not empty, it's in that namespace
2. all requests to get pipelines can specify namespace='' to query only shared pipelines
3. or specify namespace='XXX' to find pipelines in a namespace + shared pipelines
4. We need to add a config option that disables shared pipelines for organizations that definitely do not want shared pipelines
5. We need to add a KFP UI option to switch between uploading a pipeline to shared space or current namespace.

Benefits:
1. The same backend logic seems to be able to support both 'all shared' or 'all separated' use cases.
2. There's no DB migration needed, this will be a seamless transition from older versions of KFP.

Probably there are cases I haven't fully thought through. I'd suggest anyone who's willing to push this issue forward to do the following:
1. try to discuss and converge on a set of CUJs (critical user journey) that works for everyone, note we should probably start with MVPs that address immediate concerns, but make sure we can head towards the direction that all valid use-cases can be supported.
2. after consensus on the UX, discuss with maintainers a rough design of the implementation
3. contribute PRs to implement this feature in smaller self-contained steps

**Comment by Bobgy** [comment]
/cc @yanniszark 
about this

**Comment by Bobgy** [comment]
and
/cc @chensun 
who designed multi user backend separation for other KFP resources

**Comment by Jeffwan** [comment]
@Bobgy 

> We need to add a KFP UI option to switch between uploading a pipeline to shared space or current namespace.

Do we want user to upload their pipeline to share space? Should they use Managed Contributor instead like we did for experiments and runs? 

> There's no DB migration needed, this will be a seamless transition from older versions of KFP.

Pipeline schema doesn't have namespace concept. seamless migration here means implicitly migration, right?

**Comment by Bobgy** [comment]
> @Bobgy @chensun
> 
> > We need to add a KFP UI option to switch between uploading a pipeline to shared space or current namespace.
> 
> Do we want user to upload their pipeline to share space? Should they use Managed Contributor instead like we did for experiments and runs?

Some customers I worked with prefer this shared space though. A problem with manage contributor is that, they want to take a pipeline from shared space and run it in their own namespace. This will not be possible with adding contributors.

(It might be a good idea, to allow specifying pipelines in other users' namespace and KFP backend checks for that permission like contributors. Is that what you are proposing?)
 
> > There's no DB migration needed, this will be a seamless transition from older versions of KFP.
> 
> Pipeline schema doesn't have namespace concept yet. seamless migration here means implicitly migration, right? for all previous pipeline, we can think they have all empty namespace

If we do not introduce the shared pipeline concept, then we either need to figure out which pipeline belongs to which namespace or make all past pipelines inaccessible after an upgrade. This is fairly tricky to deal with.

**Comment by Jeffwan** [comment]
> A problem with manage contributor is that, they want to take a pipeline from shared space and run it in their own namespace. This will not be possible with adding contributors.

Got it, they like to "folk" the pipeline into their own namespace. Managed contributor can only manage pipeline between users but not able to public to all. I think this makes sense. 

> to allow specifying pipelines in other users' namespace
It might be hard to deal with pipeline discovery (current KFP doesn't have granular control). User may have to provide pipeline name in other user's namespace. 

I think what you propose is a simple and better way to support isolated and shared pipelines. 


> then we either need to figure out which pipeline belongs to which namespace 

I notice UI actually determine namespace from centralboard. https://github.com/kubeflow/pipelines/blob/935a9b5ba5057bc9801fee87ed17c03c2907ec85/frontend/src/lib/KubeflowClient.tsx#L23-L33. 

Do you think it's better to figure out username and corresponding namespace via headers and KFAM?
backend can parse user's email and check kfam to figure out which namespace this profile owns and the namespace other users share with the current user.  Then KFP resources can be filtered by these namespaces. 

With this we can decouple centraldashboard from KFP. I think multi-user KFP can be used separated with Central dashboard. (probably only need KFAM profile and istio)

**Comment by Bobgy** [comment]
Regarding namespace selector, I think the topic was brought up before too, if we plan to let KFP support multi user mode without central dashboard, we can build a namespace selector similar to the one in centraldashboard. It won't be so much effort on the UI side (it was designed to be very decoupled in UI code from the beginning).

Regarding to KFAM, we have decided to deprecate it, @elikatsis has sent out a PR a few days ago: https://github.com/kubeflow/pipelines/pull/4723. Let's move related discussion to that PR, and keep discussion here focused on separating pipeline resources.

**Comment by arllanos** [comment]
@Bobgy 
> 1. pipelines in DB should have a namespace column

Shall we have Namespace in pipeline_versions schema? For example I see Description is in pipelines but not in pipeline_versions. Not sure what is the criteria to decide here, but seems users are not allowed to edit or change pipelines once loaded, right?

> 4. We need to add a config option that disables shared pipelines for organizations that definitely do not want shared pipelines

A config option shall allow organizations to enable/disable the possibility pipelines can be public. Where do you think this config can be set (DB, configMap, config.json)?

**Comment by Bobgy** [comment]
> @Bobgy 
> > 1. pipelines in DB should have a namespace column
> 
> Shall we have Namespace in pipeline_versions schema? For example I see Description is in pipelines but not in pipeline_versions. Not sure what is the criteria to decide here, but seems users are not allowed to edit or change pipelines once loaded, right?

Versions are sub resources of pipelines, so they should be in the same namespace of the pipeline.

Description was more of an unfinished migration to version API, ideally we should have description field on versions too.

> > 4. We need to add a config option that disables shared pipelines for organizations that definitely do not want shared pipelines
> 
> A config option shall allow organizations to enable/disable the possibility pipelines can be public. Where do you think this config can be set (DB, configMap, config.json)?

You should use viper to get this config like other configs in API server, so that it will be configurable via config map, config.json or env var directly.

**Comment by maganaluis** [comment]
@Jeffwan @Bobgy 

We completed the implementation using KFam (for now). From a UI standpoint, we only added a checkbox to indicate if the Pipeline is "Shared" or not, Pipeline Versions of course cannot be shared. So if you select "create new pipeline..." this checkmark will not be there. 

Most of the changes were in the Backend. You can see the implementation details on this commit: https://github.com/arllanos/pipelines/commit/2c88722f6f05b67acc16c2b4e7bc54ff91c3f36c

We have not implemented an Env Variable to disable "Pipeline Sharing", however this shouldn't be too bad I think it should suffice to disable the button all together and Pipelines will be namespaced by default. 

Next week I'll switch the authentication to SubjectAccessReview and make the PR. 

One thing to note is that, "Shared Pipelines" are still deletable by other users in the System. I'm not sure what the ideal scenario should be here. 

![Screen Shot 2020-11-19 at 11 41 12 AM](https://user-images.githubusercontent.com/15258405/99703332-9af46d80-2a5c-11eb-9c3e-58654331863e.png)

**Comment by yanniszark** [comment]
@Bobdy @maganaluis thanks for your efforts on this. Some comments:

> or specify namespace='XXX' to find pipelines in a namespace + shared pipelines

@Bobgy I think it would be best to not return mixed results here. Specifying namespace='XXX' should only return pipelines from namespace 'XXX'.

> We need to add a config option that disables shared pipelines for organizations that definitely do not want shared pipelines

@Bobgy let me propose something different. Instead of turning off features, why not protect them with authorization? For example, we can have two kinds (in RBAC only, not in the MySQL DB), `Pipelines` and `ClusterPipelines` (namespace=="", name inspired from Role and ClusterRole). If an organization doesn't want to use one or the other, they simply need to not give their users permissions for one or the other.

**Comment by Bobgy** [comment]
> @bobdy @maganaluis thanks for your efforts on this. Some comments:
> 
> > or specify namespace='XXX' to find pipelines in a namespace + shared pipelines
> 
> @Bobgy I think it would be best to not return mixed results here. Specifying namespace='XXX' should only return pipelines from namespace 'XXX'.

You are right. Agree with that. So there needs to be some UI change to allow finding pipelines from either shared or namespaced ones.

> > We need to add a config option that disables shared pipelines for organizations that definitely do not want shared pipelines
> 
> @Bobgy let me propose something different. Instead of turning off features, why not protect them with authorization? For example, we can have two kinds (in RBAC only, not in the MySQL DB), `Pipelines` and `ClusterPipelines` (namespace=="", name inspired from Role and ClusterRole). If an organization doesn't want to use one or the other, they simply need to not give their users permissions for one or the other.

I guess it depends more on whether the shared pipelines will be a long term UX we will maintain. If it is, then I agree with your suggestion, that sounds like a clean solution. Do we have enough evidence it will be?

I think there're still quite some questions we need a clear answer on.
@maganaluis will you be willing to draft a rough design doc for this issue

**Comment by Bobgy** [comment]
Hmm, maybe we can start with your PR description for discussion and see if everyone can agree on that.

Note that, for above discussion, my personal opinion is that, the MVP PR do not need to implement any of the `ClusterPipeline` rbac stuff nor the "pipeline sharing" switch. We can start from the minimal and enhance based on further requests.

**Comment by Bobgy** [comment]
> One thing to note is that, "Shared Pipelines" are still deletable by other users in the System. I'm not sure what the ideal scenario should be here.

It's backward compatible behavior, so I don't think we need to worry about that right now.

**Comment by Bobgy** [comment]
We still need UI work to finish this

**Comment by Bobgy** [comment]
Remaining work item tracked in https://github.com/kubeflow/pipelines/issues/5084

**Comment by Bobgy** [comment]
To summarize on the decisions in the backend implementation when supporting namespaced pipelines: https://github.com/kubeflow/pipelines/pull/4835.

1. We now have shared pipelines and namespaced pipelines.
2. Past KFP pipeline resources are now shared pipelines, they do not belong to a namespace and all users can access them. It's TBD for implementing access control for shared pipelines, if we do in the future, we may use cluster-wide RBAC permissions.
3. Namespaced pipelines belong to a specific namespace, they are access controlled using namespaced RBAC roles/rolebindings like other KFP resources.
4. In the DB schema, both of them are pipeline resources, the difference is just whether the namespace field is empty or not.
5. All backend API behaviors are backward compatible
    * when uploading a pipeline, without namespace -> shared pipeline; with namespace -> namespaced pipeline
    * when listing pipelines, without namespace -> returns shared pipeline only; with namespace -> namespaced pipelines only
 6. To be implemented, for KFP UI integration
     * on pipeline list page, there should be a new tab or checkbox that decides whether the user is looking at shared or namespaced pipelines
     * on create pipeline page, there should be a new checkbox that decides whether the uploaded pipeline should be shared or namespaced

ref:
https://github.com/kubeflow/pipelines/pull/4835#issuecomment-770495379

**Comment by ca-scribner** [comment]
For anyone coming here and wanting a tl/dr on the issue search (some of this is a repeat of @Bobgy's great summary above): 

* [Early implementation tracking issue](https://github.com/kubeflow/pipelines/issues/1223): Includes a [design doc](https://docs.google.com/document/d/1R9bj1uI0As6umCTZ2mv_6_tjgFshIKxkSt00QLYjNV4/edit#heading=h.5s8rbufek1ax) from 2019
* [Isolation of pipeline definitions](https://github.com/kubeflow/pipelines/issues/4197) (this issue): Not fully implemented, but see summaries above.  Backend appears done (including updates to the sql DB, etc)
* [Isolation of pipeline definitions (backend)](https://github.com/kubeflow/pipelines/pull/4835): **Implemented** backend changes to support isolation of pipelines
* [Isolation of pipeline definitions (frontend)](https://github.com/kubeflow/pipelines/issues/5084): Basics spec'd out, but no work completes.  Arrikto expressed interest in doing it ~1 year ago but haven't had time (last response a few months ago).  
* [Isolation of pipeline definitions (sdk)](https://github.com/kubeflow/pipelines/issues/4197#issuecomment-953925791): As noted above the SDK needs at least some edits to enable this.  Not sure if it just needs python changes or something deeper in the golang

**Comment by Bobgy** [comment]
/cc @zijianjoy 

Re SDK changes, it's just python changes, golang clients are auto generated before each release.

**Comment by ca-scribner** [comment]
Oh great! so something like the api behind [this call](https://github.com/kubeflow/pipelines/blob/14017cced610eb289eaaba858df1183d168549ef/sdk/python/kfp/_client.py#L1299) does have the ability to accept namespace once things are regenerated?

**Comment by grobbie** [comment]
Basically for the front end iis it a matter of transplanting code like this:

https://github.com/kubeflow/pipelines/blob/0c0e4b61050e6fea6f20dd35e14eb3edc214bbba/frontend/src/pages/ExperimentList.tsx#L203-L204

and this:

https://github.com/kubeflow/pipelines/blob/0c0e4b61050e6fea6f20dd35e14eb3edc214bbba/frontend/src/pages/ExperimentList.tsx#L291-L294

Into the pipeline screens:
https://github.com/kubeflow/pipelines/blob/0c0e4b61050e6fea6f20dd35e14eb3edc214bbba/frontend/src/pages/PipelineList.tsx#L178

?

**Comment by StefanoFioravanzo** [comment]
Hi there! We (Arrikto) had promised we would work on the UI for namespaced pipelines, and we finally did :) 

We haven an internal version of the KFP UI which supports both namespaced (private) and shared (non-namespaced) pipelines, with a tab-page approach (See screenshots). We added these tabs to

- the main pipelines definitions page
- the dialog to create a new pipeline [version]
- the dialog to select a new pipeline to start a [recurring] run

While at this, we found out that:

1. the KFP Python Client lacks support for namespaced selection, and
2. the API Server is not following the correct authorization path, since the endpoint is not GCRP but an HTTP one. This causes SubjectAccessReview issues.

We fixed all these nuances and we are preparing a few PRs. We definitely want this work to land upstream as soon as possible, so that it gets into the next release.

cc @james-jwu @chensun @numerology 

I noticed that there has been some work by @grobbie . @grobbie I appreciate the work you have done so far, may I propose you hold it off for a bit and see what we have to propose?

![Screenshot 2022-03-22 at 15 50 39](https://user-images.githubusercontent.com/3354305/159510745-11fe70f0-8fdb-4afa-83e9-70111300e468.png)
![Screenshot 2022-03-22 at 15 51 05](https://user-images.githubusercontent.com/3354305/159510757-2345abda-b075-437d-a862-76e4281a1f03.png)
![Screenshot 2022-03-22 at 15 50 53](https://user-images.githubusercontent.com/3354305/159510760-d0599114-01ce-4fbc-a3e4-760a3b3f214e.png)

**Comment by chensun** [comment]
> Hi there! We (Arrikto) had promised we would work on the UI for namespaced pipelines, and we finally did :)
> 
> We haven an internal version of the KFP UI which supports both namespaced (private) and shared (non-namespaced) pipelines, with a tab-page approach (See screenshots). We added these tabs to
> 
> * the main pipelines definitions page
> * the dialog to create a new pipeline [version]
> * the dialog to select a new pipeline to start a [recurring] run
> 
> While at this, we found out that:
> 
> 1. the KFP Python Client lacks support for namespaced selection, and
> 2. the API Server is not following the correct authorization path, since the endpoint is not GCRP but an HTTP one. This causes SubjectAccessReview issues.
> 
> We fixed all these nuances and we are preparing a few PRs. We definitely want this work to land upstream as soon as possible, so that it gets into the next release.
> 
> cc @zijianjoy @james-jwu @chensun @numerology
> 
> I noticed that there has been some work by @grobbie . @grobbie I appreciate the work you have done so far, may I propose you hold it off for a bit and see what we have to propose?
> 
> ![Screenshot 2022-03-22 at 15 50 39](https://user-images.githubusercontent.com/3354305/159510745-11fe70f0-8fdb-4afa-83e9-70111300e468.png) ![Screenshot 2022-03-22 at 15 51 05](https://user-images.githubusercontent.com/3354305/159510757-2345abda-b075-437d-a862-76e4281a1f03.png) ![Screenshot 2022-03-22 at 15 50 53](https://user-images.githubusercontent.com/3354305/159510760-d0599114-01ce-4fbc-a3e4-760a3b3f214e.png)

Hi @StefanoFioravanzo, before moving on to sending PRs, can you please write a design doc for the proposal, and possibly present it in the Kubeflow Pipelines community meeting? We're going to proposal that going forward any changes that affect user experience should probably go through a design review process. Thanks!

**Comment by StefanoFioravanzo** [comment]
@chensun Yes we can do this! I see that the next Europe-friendly KFP meeting is next Wed, March 30th. We could demo our approach during this meeting, and then work on preparing the design doc. Otherwise the next available meeting is on April 27th. WDYT?

**Comment by chensun** [comment]
> @chensun Yes we can do this! I see that the next Europe-friendly KFP meeting is next Wed, March 30th. We could demo our approach during this meeting, and then work on preparing the design doc. Otherwise the next available meeting is on April 27th. WDYT?

That sounds great. Thanks!
I'm reserving a spot for you on Mar 30, but feel free to change if needed.

**Comment by juliusvonkohout** [comment]
We have an implementation that goes way beyond just namespaced pipeline definitions https://github.com/kubeflow/pipelines/pull/7406

**Comment by juliusvonkohout** [comment]
> > @chensun Yes we can do this! I see that the next Europe-friendly KFP meeting is next Wed, March 30th. We could demo our approach during this meeting, and then work on preparing the design doc. Otherwise the next available meeting is on April 27th. WDYT?
> 
> That sounds great. Thanks! I'm reserving a spot for you on Mar 30, but feel free to change if needed.

I proposed 2022-04-14
[Host] Start [recording](https://drive.google.com/file/d/11E721i9agYsU0JkQGgv76h9yIay9varC/view?usp=sharing)
[juliusvonkohout] namespace isolation for v1 artifacts and pipeline definitions https://github.com/kubeflow/pipelines/pull/7406

But Arriktos approach looks a bit more sophisticated. I hope that they also store the namespaced pipeline definitions in separate mini buckets per namespace and not the insecure default mlpipeline bucket.

**Comment by StefanoFioravanzo** [comment]
@juliusvonkohout Hi! 

> But Arriktos approach looks a bit more sophisticated. I hope that they also store the namespaced pipeline definitions in separate minio buckets per namespace and not the insecure default mlpipeline bucket.

Thanks, I do like our approach on the frontend, I think it has a better UX. Where is the design issue about how you are implementing artifact isolation?

**Comment by juliusvonkohout** [comment]
> @juliusvonkohout Hi!
> 
> > But Arriktos approach looks a bit more sophisticated. I hope that they also store the namespaced pipeline definitions in separate minio buckets per namespace and not the insecure default mlpipeline bucket.
> 
> Thanks, I do like our approach on the frontend, I think it has a better UX. Where is the design issue about how you are implementing artifact isolation?

Hello,

I do not care about the webdesign that much. If your webdesign is supported and is better that is the way to go. I give a shit about developer pride and who gets the credit, but I care about enterprise readiness and security. This is what the "just namespaced pipelines UX" will be most likely lacking.

The "design proposal" is here https://github.com/kubeflow/pipelines/pull/7219#issuecomment-1024086393 and implemented here https://github.com/kubeflow/pipelines/pull/7406 . It works in practice but we want to upstream it because of the maintenance burden.

By default the security / namespace isolation is completely broken in Kubeflow.  The mysql database stores only the metadata of the pipeline, but the actual content is on minio.

Each user has admin access (mlpipeline-minio-artifact secret) to minio and stores everything in one bucket called mlpipeline. E.g. user "development" can destroy the works of user "production". You can interrupt and destroy other users pipelines and artifacts.

I tried out multiple approaches and exchanged with other users and security guys.  The only sane and scalable approach is one Minio bucket per namespace. The bucket is only available to the user/profile with the same name. I have implemented that with automatic bucket generation, policies and per-user passwords. We also support shared pipelines, that will stay in the mlpipeline bucket. The per-namespace bucket stores pipelines and artifacts per namespace that were before insecure in the one-for-all mlpipeline bucket. 

So in the end we just had to make the remaining calls in the apiserver, persistenceagent and frontend namespace specific. I think it would be even better to just store the pipeline definitions in the mysql database instead of Minio. Nevertheless we still need to separate the pipeline artifacts per user on Minio.

The namespace contributor functionality is also broken and easily exploitable, but that is another topic for another day. And this is only for v1 pipelines. For v2 pipelines there is the next big insecurity by design at https://github.com/kubeflow/pipelines/issues/4790 . In general no new feature should be accepted that is not namespace-isolated.

What i propose is to get https://github.com/kubeflow/pipelines/pull/7406 merged and you do whatever you like with the webinterface afterwards. Alternatively we could also try to directly merge your approach with our namespace isolation branch.

**Comment by StefanoFioravanzo** [comment]
@juliusvonkohout Thanks for the explanation. I agree with you points, but I think that we should address artifact isolation and pipeline definition isolation as two distinct efforts that go in parallel. Yes the former includes the latter in some sense, because of the way pipeline definitions are stored. So let's say that on the one hand there is the artifact proposal, and on the other the API and UI work for support namespaced pipelines definitions.

I would suggest keeping these separate, this simplifies the review of the design and the implementations. I think we are missing an authoritative description of the proposed implementation in this issue https://github.com/kubeflow/pipelines/issues/4649

**Comment by juliusvonkohout** [comment]
@StefanoFioravanzo  I think there is enough discussion in all those issues regarding the different approaches. You need the separate buckets for proper pipeline definition separation. Otherwise you have to store them completely in the MySQL database.

The most important thing is to get someone with /LGTM and /approve permissions on board and merge it in time for Kubeflow 1.6. @Bobgy  told me that he has left the team and we should contact @zijianjoy and @chensun

**Comment by ca-scribner** [comment]
I'm new to this scene so not sure what the standard is, but [this](https://docs.google.com/document/d/1r7XKNncJ6du1EnX3iiUqOJXMMA2nvoXXUarOXe_sm1o/edit?resourcekey=0-pGC-6oMc2Iog5hKW1eJHAA#) was a (unrelated) design proposal that was pitched in the last meeting.  Something like that (not necessarily as detailed) that gives a snapshot of what is currently being proposed (without people having to piece together the conversation history) would make it easier to get buy-in from others.  iirc @StefanoFioravanzo just showed us the implementation in the UI as a demo rather than with a doc, so I think there's some flexibility on format

wrt the private/shared pipeline definitions (@StefanoFioravanzo et al) and the full artifact isolation (pipeline definitions, results, etc.  @juliusvonkohout et al), to me they're separate and implementing the first can be done while discussing the second **unless** there's some awkward incompatibility on the database side.  I don't think there is, but I'm not the best judge.  As long as that's ok, it seems like the feature of full isolation could be implemented on top of the pipeline isolation changes

**Comment by StefanoFioravanzo** [comment]
@ca-scribner I completely agree! There are two orthogonal discussions that can be carried on separately. We will be following up with a design doc for namespaced pipelines definitions very soon.

**Comment by juliusvonkohout** [comment]
> I'm new to this scene so not sure what the standard is, but [this](https://docs.google.com/document/d/1r7XKNncJ6du1EnX3iiUqOJXMMA2nvoXXUarOXe_sm1o/edit?resourcekey=0-pGC-6oMc2Iog5hKW1eJHAA#) was a (unrelated) design proposal that was pitched in the last meeting. Something like that (not necessarily as detailed) that gives a snapshot of what is currently being proposed (without people having to piece together the conversation history) would make it easier to get buy-in from others. iirc @StefanoFioravanzo just showed us the implementation in the UI as a demo rather than with a doc, so I think there's some flexibility on format
> 
> wrt the private/shared pipeline definitions (@StefanoFioravanzo et al) and the full artifact isolation (pipeline definitions, results, etc. @juliusvonkohout et al), to me they're separate and implementing the first can be done while discussing the second **unless** there's some awkward incompatibility on the database side. I don't think there is, but I'm not the best judge. As long as that's ok, it seems like the feature of full isolation could be implemented on top of the pipeline isolation changes

The file you linked is not publicly available. I can rewrite the design proposal and make a diagramm if that is what you need.

**Comment by ca-scribner** [comment]
Sorry, I believe it is accessible if you join the kfp Google group but I
might be wrong about that. It was a pretty standard design doc with a few
explanations and user stories, etc.

**Comment by juliusvonkohout** [comment]
Yes , I can acces it after joining the group. I will try to make a proposal during the next week.

**Comment by juliusvonkohout** [comment]
What do you think about the following design proposal https://docs.google.com/document/d/1DVHbT1RIv_VaIzWz77YCBYlBX7WEMW61AAm-ZHdWZFU/edit?usp=sharing ? Should i present it at the community call in 9 hours?

**Comment by james-jwu** [comment]
The next KFP community meeting is on 4/13 5:30PT. (The meeting you were looking at is probably the Kubeflow community meeting on 4/5)

**Comment by juliusvonkohout** [comment]
> The next KFP community meeting is on 4/13 5:30PT. (The meeting you were looking at is probably the Kubeflow community meeting on 4/5). Feel free to add an agenda item here: [http://bit.ly/kfp-meeting-notes](https://www.google.com/url?q=http://bit.ly/kfp-meeting-notes&sa=D&source=calendar&ust=1649609229214892&usg=AOvVaw1WrxaT2nTh7E2B7NsUqqR6)

I hope you can see that
![image](https://user-images.githubusercontent.com/45896133/161833438-4db7f9aa-aa09-485d-8c52-b6cfac6e8718.png)

Maybe the one day off is the European time difference.

**Comment by StefanoFioravanzo** [comment]
@chensun @james-jwu @zijianjoy Following up from our discussion about namespaced pipelines during last week's community meeting, here is a design doc that details our proposal: https://docs.google.com/document/d/1fM4y2L1IVqVj-iiNjYFRRktdCh7FQXgU2XpaYLaqt-A

Note that this doc is about supporting uploading and viewing namespaced pipeline definitions. Dealing with artifacts isolation is **not** part of this effort and is something @juliusvonkohout is working on. Please take a look and let me know if you agree with the plan! Happy to discuss and iterate.

**Comment by juliusvonkohout** [comment]
> Following up from our discussion about namespaced pipelines during last week's community meeting, here is a design doc that details our proposal: https://docs.google.com/document/d/1fM4y2L1IVqVj-iiNjYFRRktdCh7FQXgU2XpaYLaqt-A

I had to request access, are you sure that is is properly public?

**Comment by StefanoFioravanzo** [comment]
@juliusvonkohout Indeed I had pasted the wrong link, sorry for that. I amended my comment above. Here is the correct link again https://docs.google.com/document/d/1fM4y2L1IVqVj-iiNjYFRRktdCh7FQXgU2XpaYLaqt-A/edit?usp=sharing&resourcekey=0-kd5loyP7w3PBD0ug6ECmLQ

**Comment by juliusvonkohout** [comment]
> @juliusvonkohout Indeed I had pasted the wrong link, sorry for that. I amended my comment above. Here is the correct link again https://docs.google.com/document/d/1fM4y2L1IVqVj-iiNjYFRRktdCh7FQXgU2XpaYLaqt-A/edit?usp=sharing&resourcekey=0-kd5loyP7w3PBD0ug6ECmLQ

This looks very good. I hope you can merge it soon. The sooner it is merged, the sooner we can modify it for minio namespace isolation.

**Comment by StefanoFioravanzo** [comment]
@juliusvonkohout Thanks for the kind words. Let's wait for feedback from the KFP folks :)

**Comment by juliusvonkohout** [comment]
> What do you think about the following design proposal https://docs.google.com/document/d/1DVHbT1RIv_VaIzWz77YCBYlBX7WEMW61AAm-ZHdWZFU/edit?usp=sharing ? Should i present it at the community call in 9 hours?

I have tried to incorporate some answers to the questions in the community meeting and enabled commenting. You can also reach me via slack.

**Comment by juliusvonkohout** [comment]
Minio and pipeline definition per namespace isolation is planned for the future and was the main talking point of the last community meeting. James Liu @zijianjoy  wants to split it into multiple PRs. 

1. Hopefully moving pipeline definitions/templates from Minio+MySQL to MySQL only (James Liu Google) and the API change proposal https://docs.google.com/document/d/19OfU-hIsY4xBKA6b_F3dYIbSgrLMrVshucUSsyqotQ4/edit# (Diana Atanasova Vmware @difince ) are the first two important steps. I even found a quite elegant and not-invasive solution with @difince today.
2.  If this is done i will add the Minio muliple bucket logic as in https://github.com/kubeflow/pipelines/pull/7406/files#diff-0b2567187b25771c77e1223dd7ccc6068f599bf663b63423ebb33e19f6d56a32 and discard the rest of this pull request. I will also need to edit sync.py to create bucket,policies, individual users and passwords per profile/namespace, add/change the configmap artifactrepositories and kfp-launcher. I have a working solution but i need to discuss the details with @zijianjoy again
3. Arrikto (@StefanoFioravanzo ) might add the UI changes for the shared/namespaced pipeline definitions/templates too as proposed in https://github.com/kubeflow/pipelines/issues/4197 because their UI is way prettier than mine ;-). 
5. I might even add a fix for the cache-server such that deleted artifacts are not considered successfully cached anymore

In total this would yield a rather pleasant multi-user enterprise-level readiness.
It depends on the speed of all participants if we can get enterprise ready namespace isolation in upstream 1.6 or whether 1.7 is more realistic. I have the Proof of concept it working locally in my own Kubeflow instances so far.

The next big step would then to repair the insecure namespace sharing ("Manage contributors" in the UI). Currently it can be easily exploited to access other users namespaces even after they have removed you as contributor or removed you completely from the kubeflow instance (e.g. employee leaving the company). The reason is that serviceaccounts and secrets are not recreated when removing contributors.

**Comment by juliusvonkohout** [comment]
I have something elegant and mergeable ready https://github.com/kubeflow/pipelines/pull/7725
It is also quite easy to try out independently. If this is accepted i can upstream the UI changes in another merge request.

**Comment by hanakawamomo** [comment]
Hi there, are namespaced pipelines something the community still wants to support? Can we at least support the namespace param when uploading pipelines and pipeline versions in the python SDK? It would be really useful for our use case

Thanks

**Comment by juliusvonkohout** [comment]
> Hi there, are namespaced pipelines something the community still wants to support? Can we at least support the namespace param when uploading pipelines and pipeline versions in the python SDK? It would be really useful for our use case
> 
> Thanks

You need to ask @StefanoFioravanzo and attend the KFP meeting on Wednesday

**Comment by StefanoFioravanzo** [comment]
@hanakawamomo @juliusvonkohout Hi! Please see https://github.com/kubeflow/pipelines/pull/8196. The PR include the backend and client changes to support namespaced pipelines.

As soon as that one gets merges, we will open a PR that includes the UI changes described in the doc https://github.com/kubeflow/pipelines/issues/4197#issuecomment-1092211688

**Comment by elikatsis** [comment]
During the preparation of the frontend PR, which as mentioned above depends on merging #8196, we realized a better UX on a specific part.

If you take a look at the [Frontend part of the design doc](https://docs.google.com/document/d/1fM4y2L1IVqVj-iiNjYFRRktdCh7FQXgU2XpaYLaqt-A/edit?resourcekey=0-kd5loyP7w3PBD0ug6ECmLQ#heading=h.h7d4yhnc6s55) you will notice a screenshot corresponding to the "Upload pipeline or pipeline version" page.

When uploading a new pipeline version, the decision on whether it will be private or shared is implicitly taken and its dependent on the parent pipeline. Thus, the toggle `Private`/`Shared` doesn't matter and we should be hiding it. Here's how a user can reach this state:

1. Pipelines list page -> `+ Upload pipeline` button -> `Create a new pipeline version under an existing pipeline` toggle -> `Choose` pipeline
2. Pipeline details page -> `+ Upload version` (the aforementioned toggle is already selected and the pipeline is already chosen)

However, upon choosing a pipeline to upload a new version for, we can improve the UX dramatically by showing whether it's "Private" or "Shared".

To unlock this better UX, we need to be able to tell whether a pipeline belongs in a namespace or not. For this reason, we decided to also augment [the API response for a Pipeline](https://github.com/kubeflow/pipelines/blob/f25127fe769bce04be30f81b6e55ee1469705484/backend/src/apiserver/server/api_converter.go#L60-L85) to include the namespace as a resource reference, **identically to what happens with Experiments** ([source](https://github.com/kubeflow/pipelines/blob/f25127fe769bce04be30f81b6e55ee1469705484/backend/src/apiserver/server/api_converter.go#L31-L41)).

The change is pretty trivial, we have the PR on its way and I'll edit this comment to link to it.

cc @jlyaoyuli @zijianjoy @chensun @Linchin @juliusvonkohout @StefanoFioravanzo

**Comment by juliusvonkohout** [comment]
> During the preparation of the frontend PR, which as mentioned above depends on merging #8196, we realized a better UX on a specific part.
> 
> If you take a look at the [Frontend part of the design doc](https://docs.google.com/document/d/1fM4y2L1IVqVj-iiNjYFRRktdCh7FQXgU2XpaYLaqt-A/edit?resourcekey=0-kd5loyP7w3PBD0ug6ECmLQ#heading=h.h7d4yhnc6s55) you will notice a screenshot corresponding to the "Upload pipeline or pipeline version" page.
> 
> When uploading a new pipeline version, the decision on whether it will be private or shared is implicitly taken and its dependent on the parent pipeline. Thus, the toggle `Private`/`Shared` doesn't matter and we should be hiding it. Here's how a user can reach this state:
> 
> 1. Pipelines list page -> `+ Upload pipeline` button -> `Create a new pipeline version under an existing pipeline` toggle -> `Choose` pipeline
> 2. Pipeline details page -> `+ Upload version` (the aforementioned toggle is already selected and the pipeline is already chosen)
> 
> However, upon choosing a pipeline to upload a new version for, we can improve the UX dramatically by showing whether it's "Private" or "Shared".
> 
> To unlock this better UX, we need to be able to tell whether a pipeline belongs in a namespace or not. For this reason, we decided to also augment [the API response for a Pipeline](https://github.com/kubeflow/pipelines/blob/f25127fe769bce04be30f81b6e55ee1469705484/backend/src/apiserver/server/api_converter.go#L60-L85) to include the namespace as a resource reference, **identically to what happens with Experiments** ([source](https://github.com/kubeflow/pipelines/blob/f25127fe769bce04be30f81b6e55ee1469705484/backend/src/apiserver/server/api_converter.go#L31-L41)).
> 
> The change is pretty trivial, we have the PR on its way and I'll edit this comment to link to it: https://github.com/kubeflow/pipelines/pull/8375
> 
> cc @jlyaoyuli @zijianjoy @chensun @Linchin @juliusvonkohout @StefanoFioravanzo

That sounds very reasonable

**Comment by juliusvonkohout** [comment]
https://github.com/kubeflow/pipelines/pull/11965
